src/components/ItemList.svelte

```diff
@@ -20,44 +21,75 @@
 
   export let showFilter = true;
 
+  const features = new Set(
+    flatten(
+      items.map(
+        (item) =>
+          (item.features && item.features.map((f) => f.toLowerCase())) || []
+      )
+    )
+  );
   let filteredItems = items.filter((item) => !isExpired(item.expires));
   let pagedItems;
   let paginated = true;
 
   let currentPage = writable(1);
   setContext("currentPage", currentPage);
 
-  const searchText = getContext("searchText");
-  const goToPage = (page, perPage = DEFAULT_ITEMS_PER_PAGE) => {
-    pagedItems =
-      filteredItems.length > 0
-        ? chunk([...filteredItems], perPage)[page - 1]
-        : [];
-  };
-
-  const showExpired = getContext("showExpired");
-
+  // re-filter items when showExpired or search text changes
   $: {
-    if (paginated) {
-      goToPage($currentPage);
-    } else {
-      goToPage(1, filteredItems.length);
-    }
+    const search = $pageState.search || "";
+    // don't use $currentPage, because we don't want to call this reactively
+    // when paginating
+    currentPage.set(1);
+
+    const featureMatch = (item) =>
+      item.features &&
+      item.features.some((feature) =>
+        feature.toLowerCase().includes(search.toLowerCase())
+      );
+
+    // if the search string is an exact match for a particular feature,
+    // then filter just on that, otherwise filter on search text
+    filteredItems = features.has(search)
+      ? items.filter(
+          (item) =>
+            item.features &&
+            item.features.some(
+              (feature) => feature.toLowerCase() === search.toLowerCase()
+            )
+        )
+      : items.filter(
+          (item) => item.name.includes(search) || featureMatch(item)
+        );
+
+    // also filter out expired items (if we're not showing expired)
+    filteredItems = $pageState.showExpired
+      ? filteredItems
+      : filteredItems.filter((item) => !isExpired(item.expires));
+
+    // sort the filtered results so matches to features come before
+    // general matches
+    filteredItems = filteredItems.sort((item1, item2) => {
+      return (featureMatch(item2) ? 1 : 0) - (featureMatch(item1) ? 1 : 0);
+    });
   }
 
-  // re-filter items when showExpired or $searchText changes
+  // update pagination when either pagination changes or filtered list changes
+  // (above)
   $: {
-    const shownItems = $showExpired
-      ? items
-      : items.filter((item) => !isExpired(item.expires));
-    filteredItems = shownItems.filter((item) =>
-      item.name.includes($searchText)
-    );
-    // show the first page of result
-    currentPage.set(1);
-    // even if currentPage is already 1, we need to manually call goToPage() to get the first page
-    goToPage(1);
+    const perPage = paginated ? DEFAULT_ITEMS_PER_PAGE : filteredItems.length;
+    pagedItems =
+      filteredItems.length > 0
+        ? chunk([...filteredItems], perPage)[$currentPage - 1]
+        : [];
```

Comment:
Do you think we should update pageState to persist pagination state too?

Reply:
I think we probably should, yes. Let's save it for a followup though, I don't think it's too pressing. Filed #528